### PR TITLE
[Exp PyROOT] Add pythonization for TTree::Branch

### PR DESCRIPTION
**Pull request by etejedor**
This PR modifies the behaviour of `TTree::Branch` in Python so that proxy references can be passed as arguments from the Python side, more precisely in cases where the C++ implementation of the method expects the address of a pointer.

For example:
```python
v = ROOT.std.vector('int')()
t.SetBranchAddress("my_vector_branch", v)
```

Tests for the aforementioned pythonization are also included in this PR. There is one test missing, which covers the case when the new branch to be created is a field of a struct, for example:
```python
ROOT.gInterpreter.Declare("""
  struct MyStruct {
    int myint1;
    int myint2;
  };
""")

ms = ROOT.MyStruct()
ms.myint1 = some_value
t.Branch('myint1b', addressof(ms, 'myint1'), 'myint1balias/I')
```
In the current PyROOT, in order to get the address of the struct field, one would use the `AddressOf` function. The equivalent in modern Cppyy seems to be `addressof`, but it has a bug when using it as specified above. [A PR](https://bitbucket.org/wlav/cpycppyy/pull-requests/10/addressof-fix-retrieval-of-address-of-an) has been submitted to Cppyy to fix the issue.

Moreover, this PR includes an additional fix and test for `TTree::SetBranchAddress`.

### CODE REVIEW

File: bindings/pyroot_experimental/PyROOT/src/PyROOTModule.cxx [hunk 1]
```diff
@@ -31,6 +31,8 @@ static PyMethodDef gPyROOTMethods[] = {{(char *)"AddBranchAttrSyntax", (PyCFunct
                                         (char *)"Allow to access branches as tree attributes"},
                                        {(char *)"SetBranchAddressPyz", (PyCFunction)PyROOT::SetBranchAddressPyz, METH_VARARGS,
                                         (char *)"Pythonization for TTree::SetBranchAddress"},
+                                       {(char *)"BranchPyz", (PyCFunction)PyROOT::BranchPyz, METH_VARARGS,
```
Comment (dpiparo) [inline bindings/pyroot_experimental/PyROOT/src/PyROOTModule.cxx]: The comment is fine but perhaps we can choose something which is less generic than "Pythonization for TTree::Branch?"

File: bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx [hunk 2]
```diff
@@ -236,3 +236,155 @@ PyObject *PyROOT::SetBranchAddressPyz(PyObject * /* self */, PyObject *args)
    // Not the overload we wanted to pythonize, return None
    Py_RETURN_NONE;
 }
+
+// Try ( const char*, void*, const char*, Int_t = 32000 )
+PyObject *TryBranchLeafListOverload(int argc, PyObject *args)
+{
+   PyObject *treeObj = nullptr;
+   PyObject *name = nullptr, *address = nullptr, *leaflist = nullptr, *bufsize = nullptr;
```
Comment (dpiparo) [inline bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx]: why 2 lines?

File: bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx [hunk 3]
```diff
@@ -236,3 +236,155 @@ PyObject *PyROOT::SetBranchAddressPyz(PyObject * /* self */, PyObject *args)
    // Not the overload we wanted to pythonize, return None
    Py_RETURN_NONE;
 }
+
+// Try ( const char*, void*, const char*, Int_t = 32000 )
```
Comment (dpiparo) [inline bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx]: Perhaps for our future selves we can comment a bit better the function with doxy?

File: bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx [hunk 4]
```diff
@@ -236,3 +236,155 @@ PyObject *PyROOT::SetBranchAddressPyz(PyObject * /* self */, PyObject *args)
...
+
+   if (PyArg_ParseTuple(args, const_cast<char *>("OO!OO!|O!:Branch"),
+                        &treeObj,
+                        &CPyCppyy_PyUnicode_Type, &name,
+                        &address,
+                        &CPyCppyy_PyUnicode_Type, &leaflist,
+                        &PyInt_Type, &bufsize)) {
+
+      auto treeProxy = (CPPInstance *)treeObj;
+      TTree *tree = (TTree *)GetClass(treeProxy)->DynamicCast(TTree::Class(), treeProxy->GetObject());
```
Comment (dpiparo) [inline bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx]: auto
```suggestion
      auto tree = (TTree *)GetClass(treeProxy)->DynamicCast(TTree::Class(), treeProxy->GetObject());
```

File: bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx [hunk 5]
```diff
@@ -236,3 +236,155 @@ PyObject *PyROOT::SetBranchAddressPyz(PyObject * /* self */, PyObject *args)
...
+      }
+   }
+   PyErr_Clear();
+
+   Py_RETURN_NONE;
+}
+
+// Try ( const char*, const char*, T**, Int_t = 32000, Int_t = 99 )
+// or  ( const char*,              T**, Int_t = 32000, Int_t = 99 )
+PyObject *TryBranchPtrToPtrOverloads(int argc, PyObject *args)
```
Comment (dpiparo) [inline bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx]: ditto

File: bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx [hunk 6]
```diff
@@ -236,3 +236,155 @@ PyObject *PyROOT::SetBranchAddressPyz(PyObject * /* self */, PyObject *args)
...
+{
+   PyObject *treeObj = nullptr;
+   PyObject *name = nullptr, *clName = nullptr, *address = nullptr, *bufsize = nullptr, *splitlevel = nullptr;
```
Comment (dpiparo) [inline bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx]: ditto

File: bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx [hunk 7]
```diff
@@ -236,3 +236,155 @@ PyObject *PyROOT::SetBranchAddressPyz(PyObject * /* self */, PyObject *args)
...
+                           &PyInt_Type, &bufsize,
+                           &PyInt_Type, &splitlevel)) {
+         bIsMatch = true;
+      } else
+         PyErr_Clear();
+   }
+
+   if (bIsMatch) {
+      auto treeProxy = (CPPInstance *)treeObj;
+      TTree *tree = (TTree *)GetClass(treeProxy)->DynamicCast(TTree::Class(), treeProxy->GetObject());
```
Comment (dpiparo) [inline bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx]:
```suggestion
      auto tree = (TTree *)GetClass(treeProxy)->DynamicCast(TTree::Class(), treeProxy->GetObject());
```

File: bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx [hunk 8]
```diff
@@ -236,3 +236,155 @@ PyObject *PyROOT::SetBranchAddressPyz(PyObject * /* self */, PyObject *args)
...
+   } else {
+      PyErr_Clear();
+      if (PyArg_ParseTuple(args, const_cast<char *>("OO!O|O!O!"),
+                           &treeObj,
+                           &CPyCppyy_PyUnicode_Type, &name,
+                           &address,
+                           &PyInt_Type, &bufsize,
+                           &PyInt_Type, &splitlevel)) {
+         bIsMatch = true;
+      } else
```
Comment (dpiparo) [inline bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx]: is it good for our coding conventions to skip braces?

File: bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx [hunk 9]
```diff
@@ -236,3 +236,155 @@ PyObject *PyROOT::SetBranchAddressPyz(PyObject * /* self */, PyObject *args)
...
+         if (((CPPInstance *)address)->fFlags & CPPInstance::kIsReference)
+            buf = (void *)((CPPInstance *)address)->fObject;
+         else
+            buf = (void *)&((CPPInstance *)address)->fObject;
+
+         if (!clName) {
+            klName = GetClass((CPPInstance *)address)->GetName();
+            argc += 1;
+         }
+      } else
```
Comment (dpiparo) [inline bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx]: is it good for our coding conventions to skip braces?

File: bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx [hunk 10]
```diff
@@ -236,3 +236,155 @@ PyObject *PyROOT::SetBranchAddressPyz(PyObject * /* self */, PyObject *args)
...
+         Utility::GetBuffer(address, '*', 1, buf, false);
+
+      if (buf && !klName.empty()) {
+         TBranch *branch = 0;
```
Comment (dpiparo) [inline bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx]:
```suggestion
         TBranch *branch = nullptr;
```

File: bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx [hunk 11]
```diff
@@ -236,3 +236,155 @@ PyObject *PyROOT::SetBranchAddressPyz(PyObject * /* self */, PyObject *args)
...
+/// implementation of the method expects the address of a pointer.
+///
+/// For example:
+/// ~~~{.python}
+/// v = ROOT.std.vector('int')()
+/// t.Branch('my_vector_branch', v)
+/// ~~~
+PyObject *PyROOT::BranchPyz(PyObject * /* self */, PyObject *args)
+{
+   // Acceptable signatures:
```
Comment (dpiparo) [inline bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx]: perhaps this is better to put in the doxy?

File: bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx (hunk 2 above)
Comment (etejedor) [reply bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx]: The first line is the tree object on which `Branch` is invoked, the second line contains the possible parameters of `Branch` in the order they appear in the signature. Seemed cleaner to me than just all in one line.

File: bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx (hunk 3 above)
Comment (etejedor) [reply bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx]: Done!

File: bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx (hunk 4 above)
Comment (etejedor) [reply bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx]: Done!

File: bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx (hunk 5 above)
Comment (etejedor) [reply bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx]: Done!

File: bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx (hunk 8 above)
Comment (etejedor) [reply bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx]: Done!

File: bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx (hunk 10 above)
Comment (etejedor) [reply bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx]: Done!

File: bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx (hunk 11 above)
Comment (etejedor) [reply bindings/pyroot_experimental/PyROOT/src/TTreePyz.cxx]: Yes, definitely

**Pull request merged**